File: 992_subarrays_with_k_different_integers.py
'''
Given an array A of positive integers, call a (contiguous, not necessarily distinct) subarray of A good if the number of different integers in that subarray is exactly K.

(For example, [1,2,3,1,2] has 3 different integers: 1, 2, and 3.)

Return the number of good subarrays of A.

 

Example 1:

Input: A = [1,2,1,2,3], K = 2
Output: 7
Explanation: Subarrays formed with exactly 2 different integers: [1,2], [2,1], [1,2], [2,3], [1,2,1], [2,1,2], [1,2,1,2].
Example 2:

Input: A = [1,2,1,3,4], K = 3
Output: 3
Explanation: Subarrays formed with exactly 3 different integers: [1,2,1,3], [2,1,3], [1,3,4].
 

Note:

1 <= A.length <= 20000
1 <= A[i] <= A.length
1 <= K <= A.length
'''

from collections import Counter

class Solution:
    def subarraysWithKDistinct(self, A: List[int], K: int) -> int:
        return self.atMostK(A, K) - self.atMostK(A, K-1)
    
    def atMostK(self, A, K):
        count = Counter()
        res = s = 0
        for e in range(len(A)):
            if count[A[e]] == 0:
                K -= 1
            count[A[e]] += 1
            while K < 0:
                count[A[s]] -= 1
                if count[A[s]] == 0:
                    K += 1
                s += 1
            res += (e - s + 1) 
        
        return res
                

File: problems/992_subarrays_with_k_different_integers.java
/*
Given an integer array nums and an integer k, return the number of good subarrays of nums.

A good array is an array where the number of different integers in that array is exactly k.

For example, [1,2,3,1,2] has 3 different integers: 1, 2, and 3.
A subarray is a contiguous part of an array.

 

Example 1:

Input: nums = [1,2,1,2,3], k = 2
Output: 7
Explanation: Subarrays formed with exactly 2 different integers: [1,2], [2,1], [1,2], [2,3], [1,2,1], [2,1,2], [1,2,1,2]
Example 2:

Input: nums = [1,2,1,3,4], k = 3
Output: 3
Explanation: Subarrays formed with exactly 3 different integers: [1,2,1,3], [2,1,3], [1,3,4].
 

Constraints:

1 <= nums.length <= 2 * 104
1 <= nums[i], k <= nums.length
*/

class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        int[] distinctCount = new int[nums.length + 1];
        int totalCount = 0;
        int left = 0;
        int right = 0;
        int currCount = 0;
        while (right < nums.length) {
            if (distinctCount[nums[right++]]++ == 0) {
                k--;
            }
            if (k < 0) {
                --distinctCount[nums[left++]];
                k++;
                currCount = 0;
            }
            if (k == 0) {
                while (distinctCount[nums[left]] > 1) {
                    --distinctCount[nums[left++]];
                    currCount++;
                }
                totalCount += (currCount + 1);
            }
        }
        return totalCount;
    }
}

